File: 3/3.cpp
#include <iostream>
#include <cmath>
using namespace std;


bool is_prime(long long n) {
	for (long long i=2; i <= ceil(sqrt(n)); i++) {
		if (n%i == 0) return false;
	}
	return true;
}


long long biggest_prime_factor(long long n) {
	for (long long i=ceil(sqrt(n)); i >= 2; i--) {
		if (n%i == 0 and is_prime(i))
			return i;
	} 
	return n;
}


int main() {
	unsigned long long n = 600851475143;
	cout << biggest_prime_factor(n) << endl;
	return 0;
}

File: 3/3.py
from math import sqrt, ceil

def is_prime(n: int) -> bool:
	if n == 1: return False
	if n == 2: return True

	for i in range(2, ceil(sqrt(n))+1):
		if n%i == 0:
			return False

	return True


def biggest_prime_factor(n: int) -> int:
	if is_prime(n): return n

	for i in range(ceil(sqrt(n)), 2, -1):
		if n%i: continue
		if is_prime(i): return i


n: int = 600851475143

print(biggest_prime_factor(n))